packages/bff/src/services/agreementService.ts

```diff
@@ -57,3 +149,279 @@ export const getLatestAgreement = async (
+    (x): x is string => x !== undefined
+  );
+}
+
+async function parallelGet(
+  agreement: agreementApi.Agreement,
+  { tenantProcessClient, catalogProcessClient }: PagoPAInteropBeClients,
+  { headers }: WithLogger<BffAppContext>
+): Promise<{
+  consumer: tenantApi.Tenant;
+  producer: tenantApi.Tenant;
+  eservice: catalogApi.EService;
+}> {
+  const consumerTask = tenantProcessClient.tenant.getTenant({
+    params: { id: agreement.consumerId },
+    headers,
+  });
+
+  const producerTask = tenantProcessClient.tenant.getTenant({
+    params: { id: agreement.producerId },
+    headers,
+  });
+  const eserviceTask = catalogProcessClient.getEServiceById({
+    params: { eServiceId: agreement.eserviceId },
+    headers,
+  });
+  const [consumer, producer, eservice] = await Promise.all([
+    consumerTask,
+    producerTask,
+    eserviceTask,
+  ]);
+
+  return {
+    consumer,
+    producer,
+    eservice,
+  };
+}
+
+async function getBulkAttributes(
+  ids: string[],
+  attributeProcess: PagoPAInteropBeClients["attributeProcessClient"],
+  { headers }: WithLogger<BffAppContext>
+): Promise<attributeRegistryApi.Attribute[]> {
+  async function getAttributesFrom(
+    offset: number
+  ): Promise<attributeRegistryApi.Attribute[]> {
+    const response = await attributeProcess.getBulkedAttributes(ids, {
+      queries: { offset, limit: 50 },
+      headers,
+    });
+    return response.results;
+  }
+
+  async function aggregate(
+    start: number,
+    attributes: attributeRegistryApi.Attribute[]
+  ): Promise<attributeRegistryApi.Attribute[]> {
+    const attrs = await getAttributesFrom(start);
+    if (attrs.length < 50) {
+      return attributes.concat(attrs);
+    } else {
+      return aggregate(start + 50, attributes.concat(attrs));
+    }
+  }
+
+  return aggregate(0, []);
+}
+
+function filterAttributes(
+  attributes: attributeRegistryApi.Attribute[],
+  filterIds: string[]
+): attributeRegistryApi.Attribute[] {
+  return attributes.filter((attr) => filterIds.includes(attr.id));
+}
+
+function enhanceTenantAttributes(
+  tenantAttributes: tenantApi.TenantAttribute[],
+  registryAttributes: attributeRegistryApi.Attribute[]
+): bffApi.TenantAttributes {
+  const registryAttributesMap: Map<string, bffApi.Attribute> = new Map(
+    registryAttributes.map((attribute) => [attribute.id, attribute])
+  );
+
+  const declared = tenantAttributes
+    .map((attr) => toApiDeclaredTenantAttribute(attr, registryAttributesMap))
+    .filter((x): x is bffApi.DeclaredTenantAttribute => x !== null);
+
+  const certified = tenantAttributes
+    .map((attr) => toApiCertifiedTenantAttribute(attr, registryAttributesMap))
+    .filter((x): x is bffApi.CertifiedTenantAttribute => x !== null);
+
+  const verified = tenantAttributes
+    .map((attr) => toApiVerifiedTenantAttribute(attr, registryAttributesMap))
+    .filter((x): x is bffApi.VerifiedTenantAttribute => x !== null);
+
+  return {
+    certified,
+    declared,
+    verified,
+  };
+}
```

Comment:
this utils function it's really common between Bff's services (agreement ,catalog, etc...) , I suggest to move it  a dedicated file into `utilities` folder.